src/zcash/JoinSplit.cpp

```diff
@@ -1,213 +1,212 @@
-        }
+namespace libzcash
+{
 
-        uint64_t lhs_value = vpub_old;
-        uint64_t rhs_value = vpub_new;
-
-        for (size_t i = 0; i < NumInputs; i++) {
-            // Sanity checks of input
-            {
-                // If note has nonzero value
-                if (inputs[i].note.value() != 0) {
-                    // The witness root must equal the input root.
-                    if (inputs[i].witness.root() != rt) {
-                        throw std::invalid_argument("joinsplit not anchored to the correct root");
-                    }
-
-                    // The tree must witness the correct element
-                    if (inputs[i].note.cm() != inputs[i].witness.element()) {
-                        throw std::invalid_argument("witness of wrong element for joinsplit input");
-                    }
-                }
+template<size_t NumInputs, size_t NumOutputs>
+SproutProof JoinSplit<NumInputs, NumOutputs>::prove(
+    const std::array<JSInput, NumInputs>& inputs,
+    const std::array<JSOutput, NumOutputs>& outputs,
+    std::array<SproutNote, NumOutputs>& out_notes,
+    std::array<ZCNoteEncryption::Ciphertext, NumOutputs>& out_ciphertexts,
+    uint256& out_ephemeralKey,
+    const ed25519::VerificationKey& joinSplitPubKey,
+    uint256& out_randomSeed,
+    std::array<uint256, NumInputs>& out_macs,
+    std::array<uint256, NumInputs>& out_nullifiers,
+    std::array<uint256, NumOutputs>& out_commitments,
+    uint64_t vpub_old,
+    uint64_t vpub_new,
+    const uint256& rt,
+    bool computeProof,
+    uint256* out_esk // Payment disclosure
+)
+{
+    if (vpub_old > MAX_MONEY) {
+        throw std::invalid_argument("nonsensical vpub_old value");
+    }
 
-                // Ensure we have the key to this note.
-                if (inputs[i].note.a_pk != inputs[i].key.address().a_pk) {
-                    throw std::invalid_argument("input note not authorized to spend with given key");
+    if (vpub_new > MAX_MONEY) {
+        throw std::invalid_argument("nonsensical vpub_new value");
+    }
+
+    uint64_t lhs_value = vpub_old;
+    uint64_t rhs_value = vpub_new;
+
+    for (size_t i = 0; i < NumInputs; i++) {
+        // Sanity checks of input
+        {
+            // If note has nonzero value
+            if (inputs[i].note.value() != 0) {
+                // The witness root must equal the input root.
+                if (inputs[i].witness.root() != rt) {
+                    throw std::invalid_argument("joinsplit not anchored to the correct root");
                 }
 
-                // Balance must be sensical
-                if (inputs[i].note.value() > MAX_MONEY) {
-                    throw std::invalid_argument("nonsensical input note value");
+                // The tree must witness the correct element
+                if (inputs[i].note.cm() != inputs[i].witness.element()) {
+                    throw std::invalid_argument("witness of wrong element for joinsplit input");
```

Comment:
See how the diff got completely out-of-sync here because of the header changes? Blech.